### PR TITLE
Revert changing currency to be count_by_charges

### PR DESCRIPTION
**Pull request by anothersimulacrum**
#### Summary
None

#### Purpose of change
Move the discussion from https://github.com/CleverRaven/Cataclysm-DDA/pull/51077#issuecomment-916061519 somewhere more visible.

#### Describe the solution
Change currencies from `AMMO` back to `GENERIC`.

#### Testing
Once the discussion goes somewhere, this will either be closed or I'll test it.

#### Additional context
https://github.com/CleverRaven/Cataclysm-DDA/pull/48525
CC @Zireael07 @chaohedgehogs @Entity-y

### CODE REVIEW

**Comment by Entity-y** [comment]
Instead of taking a step back and changing a solution that works, lets discuss a fix for the issue.

What about implementing a pseudo nested container for stacking individual objects, the lag problem with moving tens of items isn't present when using a container and there are no restrictions for having a container inside of a container. It could work behind the scenes and would automatically stack currency into it's nested form.

**Comment by anothersimulacrum** [comment]
The current situation is a hack, not a solution that works. That also sounds like a hack to me.

**Comment by Entity-y** [comment]
> The current situation is a hack, not a solution that works.

The current implementation resolves two issues, hack or not. Reverting back to generic could be done at any time easily, why not wait until there is working system in place before making the change back.

> The 'lag' from moving objects I am fairly sure is just because each item is handled individually

What about the issue of having several pages worth of inventory that's just currency?

**Comment by Zireael07** [comment]
@anothersimulacrum: #48937 needs a fix first **before** this is merged. I often play Cata on low powered devices and while it was never a speed demon, the issue makes the game unplayable past early game (once you've explored an area, there's almost certainly A LOT of a) shards b) rocks c) cash around)

**Comment by anothersimulacrum** [comment]
Hacks mean we don't solve issues, we let them sit because they aren't enough effort to change. We shouldn't institute hacks so we can ignore problems.

Do you mean in NPC trading? Yes, that sorely needs improvement - let's not pave over it and actually make it better.

> @anothersimulacrum: #48937 needs a fix first before this is merged. I often play Cata on low powered devices and while it was never a speed demon, the issue makes the game unplayable past early game (once you've explored an area, there's almost certainly A LOT of a) shards b) rocks c) cash around)

I agree it needs a fix, I disagree we should leave a hack in place and act as though that has fixed it.

**Comment by Zireael07** [comment]
The hack means the game is playable for people on less powerful devices. Without it, it's not playable, simple as that.

The hack has to stay in as long as needed for a proper solution to be made and then can be reverted.

**Comment by anothersimulacrum** [comment]
Can you provide some tests cases for me to profile to examine this situation?
We are both not operating on data, and that will get us nowhere.

**Comment by Zireael07** [comment]
Unfortunately one of the devices I was seeing issues on is not mine and I'm not anywhere close to it, as in it's 150 km away xD, and the other is an Android phone so I have close to zero idea on how to get you performance data, as I'm not too sure on how to install a specified commit/build as opposed to "latest stable" or "latest experimental"

**Comment by anothersimulacrum** [comment]
I want to profile on my devices, I just need cases.

**Comment by Zireael07** [comment]
@anothersimulacrum: Get a 150-200 sized stack of bones or shards or cash. Bones are probably the easiest, since skeletons drop lots. Or you can use debug menu.

Try picking them up or moving them between two containers. Dropping is also affected, although slightly less so I believe.

**Comment by anothersimulacrum** [comment]
Well, didn't even need to compile on a low-end device for this (about 40s to pick up 400 notes, as compared with ~1s to pick up 1 note with 400 charges).
image

**Comment by Entity-y** [comment]
That's real time or in game time? If it's in game time it would be a bit nonsense picking up 400 notes in one second, perhaps there needs to be some sort of middle ground.

**Comment by anothersimulacrum** [comment]
Real time.

**Comment by mqrause** [comment]
> Well, didn't even need to compile on a low-end device for this (about 40s to pick up 400 notes, as compared with ~1s to pick up 1 note with 400 charges).

The worst part about that is that it iterates over all items, so with every item picked up, it takes longer.

**Comment by anothersimulacrum** [comment]
This is stale. It needs looking at, but I won't press it further.

**Comment by kevingranade** [comment]
Deeper profiling
Screenshot from 2021-12-31 16-43-12
tl;dr Character::drop_invalid_inventory() scaling
